### PR TITLE
Implement Actions API.

### LINKED ISSUE #458: Add support for GitHub Actions
Hi,

Actions API: https://docs.github.com/en/rest/reference/actions
I am using this library for pull requests and comments, but now I also need to trigger workflows using "workflow dispatch event".

Should actions API be added to this library?
If so, can I help with implementation? Looks like no PRs have been merged lately, so I'm asking for guidance before sending any PRs.

### PR DESCRIPTION
**Pull request by K0Te**
Closes #458 

Hi @phadej ,

I've implemented just ~5% of Actions API at the moment. I have many questions and need your guidance before proceeding further:

- GitHub has changed documentation URLs, I am using new ones in the PR. Is it ok? It's probably worth changing all URLs in the repo, as GitHub has broken redirect a little bit, for example, https://developer.github.com/v3/pulls/#get-a-single-pull-request is forwarded to https://docs.github.com/en/rest/reference/pulls#get-a-single-pull-request instead of https://docs.github.com/en/rest/reference/pulls#get-a-pull-request . This should probably be a separate issue/PR.
- Response style is different for actions API, JSON includes tagged item list and a total count:
``` json
{
  "artifacts": [
   ...
  ],
  "total_count": 13676
}
```
I see three ways of handling this:
1. A separate Response type for each list endpoint, it's simple but need  a lot of code:
``` haskell
data ArtifactList = ArtifactList
    { artifactListArtifacts :: !(Vector Artifact)
    , artifactListTotalCount :: !Int
    }
    deriving (Show, Data, Typeable, Eq, Ord, Generic
```
2.  Tag as a type-level symbol, this looks nicer, but needs extensions and moves tag towards `Endpoints` instead of `Data`:
``` haskell
data PaginatedWithTotalCount a (tag :: Symbol) = PaginatedWithTotalCount
    { paginatedWithTotalCountItems :: !(Vector a)
    , paginatedWithTotalCountTotalCount :: !Int
    }
```
3. Tag defined in instance declaration for each paginated type. This approach seems best, but might break if GitHub API uses different tags for the same type - in this case, the library would need new types, which will be confusing...
``` haskell
data WithTotalCount a = WithTotalCount
    { withTotalCountItems :: !(Vector a)
    , withTotalCountTotalCount :: !Int
    } deriving (Show, Data, Typeable, Eq, Ord, Generic)

instance Semigroup (WithTotalCount a) where
    (WithTotalCount items1 count1) <> (WithTotalCount items2 _) =
        WithTotalCount (items1 <> items2) count1

instance Foldable WithTotalCount where
    foldMap f (WithTotalCount items _) = foldMap f items

instance FromJSON (WithTotalCount Artifact) where
    parseJSON = withObject "ArtifactList" $ \o -> WithTotalCount
        <$> o .: "artifacts"
        <*> o .: "total_count
```

- Tests. I have implemented tests for reading the artifact list. Downloading is more complex - it needs workflow permissions and does not work with an empty token. Even if https://github.com/phadej/github contained actions with artifacts, other users won't be able to run integration tests with download and deletion. Should I test only API with minimal required permissions?

### CODE REVIEW

File: src/GitHub/Data/Actions.hs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    }
+  deriving (Show, Data, Typeable, Eq, Ord, Generic)
+
+data ArtifactList = ArtifactList
+    { artifactListArtifacts :: !(Vector Artifact)
+    , artifactListTotalCount :: !Int
+    }
+    deriving (Show, Data, Typeable, Eq, Ord, Generic)
+
+data PaginatedWithTotalCount a (tag :: Symbol) = PaginatedWithTotalCount
```
Comment (phadej) [inline src/GitHub/Data/Actions.hs:47]: What are these? Is artifacts pagination different then for everything else?

Comment (K0Te) [reply src/GitHub/Data/Actions.hs:47]: Yes, please check my comment above. At least I didn't find existing implementation for this pagination style.

**Comment by K0Te** [comment]
Hi @andreasabel ,

I got back to this issue and hope to implement full support for `GitHub Actions`.
Hope you could review this PR once it's ready and help to merge it.
There is one more interesting issue with the `Actions` API - https://docs.github.com/en/rest/actions/cache#delete-github-actions-caches-for-a-repository-using-a-cache-key defines a DELETE request with query parameters but at the moment only `Queries` support query parameters. Adding query parameters to `command` or `Command` would require changing lots of code, could there be a better approach?

**Comment by andreasabel** [comment]
Hi @K0Te, good to have you back on track!

> Hope you could review this PR once it's ready and help to merge it.

I'll do my best.  I am not really an expert on this package.
Make sure your functionality is covered by tests or examples, remember to document, and adhere to the current style of the API, please.  

> There is one more interesting issue ... at the moment only `Queries` support query parameters. 
> Adding query parameters to `command` or `Command` would require changing lots of code, could there be a better approach?

I suppose you are referring to this code:
https://github.com/haskell-github/github/blob/01c115466d0fe0b879dd16d85d3a730a4142eaae/src/GitHub/Data/Request.hs#L155-L164
https://github.com/haskell-github/github/blob/01c115466d0fe0b879dd16d85d3a730a4142eaae/src/GitHub/Data/Request.hs#L174-L181

To keep API breakage at bay, I'd add it to `Command` but keep the smart constructor `command` as-is.  Add a new smart constructor (e.g. `paramCommand`) that exposes the full feature set of `Command`.

**Comment by andreasabel** [comment]
I resolved the conflicts with `master` so that CI can run.

**Comment by K0Te** [comment]
Hi @andreasabel ,

This PR is finally ready for review! 🚀 
I have tested all endpoints added in the scope of this PR, everything seems to work fine on my test repo.
Some GitHub Actions API features are missing, so implementation does not cover all available API. However, I think that is should be enough for most use cases. Missing endpoints:

- https://docs.github.com/en/rest/actions/self-hosted-runner-groups
- https://docs.github.com/en/rest/actions/self-hosted-runners
- https://docs.github.com/en/rest/actions/permissions
- Enterprise-level endpoints for Cache
- https://docs.github.com/en/rest/actions/cache#delete-github-actions-caches-for-a-repository-using-a-cache-key - see discussions above, cache can be deleted by Id using an alternative endpoint

**Comment by K0Te** [comment]
@andreasabel reminding about this PR ^_^

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks for the PR! and sorry for the delayed review.

I had some stylistic requests but fixed them myself.

What remains is to document the `WithTotalCount` type.

File: github.cabal [hunk 2]
```diff
@@ -199,9 +212,9 @@ library
 
   if flag(openssl)
     build-depends:
-        HsOpenSSL              >=0.11.4.16 && <0.12
-      , HsOpenSSL-x509-system  >=0.1.0.3   && <0.2
-      , http-client-openssl    >=0.2.2.0   && <0.4
+        HsOpenSSL              >=0.11.4.16 && <1.12
+      , HsOpenSSL-x509-system  >=0.1.0.3   && <1.2
+      , http-client-openssl    >=0.2.2.0   && <1.4
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply github.cabal]: Reverted

File: spec/GitHub/Actions/ArtifactsSpec.hs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import           Data.Foldable      (for_)
+import           Data.String        (fromString)
+import qualified Data.Vector        as V
+import           System.Environment (lookupEnv)
+import           Test.Hspec
+                 (Spec, describe, it, pendingWith, shouldBe, shouldSatisfy)
+
+fromRightS :: Show a => Either a b -> b
+fromRightS (Right b) = b
+fromRightS (Left a) = error $ "Expected a Right and got a Left" ++ show a
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply spec/GitHub/Actions/ArtifactsSpec.hs:23]: No action required now.

File: src/GitHub/Data/Actions/Common.hs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+-- Common
+-------------------------------------------------------------------------------
+
+data WithTotalCount a = WithTotalCount
+    { withTotalCountItems :: !(Vector a)
+    , withTotalCountTotalCount :: !Int
+    } deriving (Show, Data, Typeable, Eq, Ord, Generic)
+
+instance Semigroup (WithTotalCount a) where
+    (WithTotalCount items1 count1) <> (WithTotalCount items2 _) =
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply src/GitHub/Data/Actions/Common.hs]: Done.

File: src/GitHub/Data/Options.hs [hunk 5]
```diff
@@ -640,3 +678,234 @@ optionsNoAssignee = IssueRepoMod $ \opts ->
...
+    , mk "head_sha" <$> headSha'
+    ]
+  where
+    mk k v = (k, Just v)
+    actor' = fmap TE.encodeUtf8 actor
+    branch' = fmap TE.encodeUtf8 branch
+    event' = fmap TE.encodeUtf8 event
+    status' = fmap TE.encodeUtf8 status
+    created' = fmap TE.encodeUtf8 created
+    headSha' = fmap TE.encodeUtf8 headSha
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply src/GitHub/Data/Options.hs:908]: No action needed.

**Comment by andreasabel** [comment]
Candidate for 0.29 at: https://hackage.haskell.org/package/github-0.29/candidate

File: src/GitHub/Data/Actions/Common.hs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import GitHub.Internal.Prelude
+import Prelude ()
+
+-------------------------------------------------------------------------------
+-- Common
+-------------------------------------------------------------------------------
+
+data WithTotalCount a = WithTotalCount
+    { withTotalCountItems      :: !(Vector a)
+    , withTotalCountTotalCount :: !Int
```
[the remark this answers is not in the document]
Comment (andreasabel) [reply src/GitHub/Data/Actions/Common.hs:20]: I did it myself, based on the comments in this PR.

**Review by andreasabel — APPROVED** [review]
Addressed the last bits.

Good to go!

**Comment by andreasabel** [comment]
Published at https://hackage.haskell.org/package/github-0.29

Big thanks, @K0Te !

**Pull request merged**